File: Algorithms/1013.partition-array-into-three-parts-with-equal-sum/partition-array-into-three-parts-with-equal-sum.go
func canThreePartsEqualSum(A []int) bool {
    //双指针法
    length := len(A)
    left,right := 0,length-1
    leftSum,rightSum,sum := A[0],A[length-1],0
    for i:=0;i<len(A);i++ {
        sum += A[i]
    }
    for (left+1)<right {
        if (leftSum == rightSum) && (rightSum * 3 == sum) {
            return true
        }
        if (leftSum != sum/3) {
            left=left+1
            leftSum += A[left]
        }
        if (rightSum != sum/3) {
            right=right-1
            rightSum += A[right]
        }
    }
    return false
}

File: Algorithms/1013.partition-array-into-three-parts-with-equal-sum/partition-array-into-three-parts-with-equal-sum.cpp
class Solution {
public:
    bool canThreePartsEqualSum(vector<int>& A) {
        int sum = accumulate(A.begin(), A.end(), 0);  //累加求和
        if (sum % 3 != 0) {
            return false; //如果和取3的余不为0，返回false
        }
        int count = 0, subSum = 0;
        for (int i = 0; i < A.size(); i ++) {
            subSum += A[i];
            if (subSum == sum / 3) {
                count ++;
                subSum = 0;
            }
            if (count == 3) {
                return true;
            }
        }
        return false;
    }
};
